Add SSH Microagent (#8436)
All-Hands-AI/OpenHands: microagents/ssh.md

```bash
scp /path/to/local/file username@hostname:/path/to/remote/directory/
```

Copy a file from the remote server:
```bash
scp username@hostname:/path/to/remote/file /path/to/local/directory/
```

Copy a directory recursively:
```bash
scp -r /path/to/local/directory username@hostname:/path/to/remote/directory/
```

## SSH Agent

Start the SSH agent:
```bash
eval "$(ssh-agent -s)"
```

Add a key to the agent:
```bash
ssh-add ~/.ssh/key_name
```

## Troubleshooting

- Check SSH service status on remote: `systemctl status sshd`
- Verify SSH port is open: `nc -zv hostname 22`
- Debug connection issues: `ssh -vvv username@hostname`
- Check permissions: SSH private keys should have 600 permissions (`chmod 600 ~/.ssh/key_name`)
- Verify known_hosts: If host key changed, remove the old entry with `ssh-keygen -R hostname`

## Secure SSH Key Management

### Local Storage with Proper Permissions

The most basic approach is to ensure proper file permissions:

```bash
# Set correct permissions for private keys
chmod 600 ~/.ssh/id_ed25519
# Set correct permissions for public keys
chmod 644 ~/.ssh/id_ed25519.pub
# Set correct permissions for SSH directory
chmod 700 ~/.ssh
```

> ### Using Password and Secrets Managers
> 
> Many solutions support storing SSH keys securely:
> - Password managers: 1Password, LastPass, Bitwarden, KeePass
> - Enterprise secrets managers: HashiCorp Vault, Azure Key Vault, AWS Secrets Manager
> 
> ## Security Best Practices
> 
> - Use key-based authentication instead of passwords
> - Use strong, unique keys (ED25519 or RSA with at least 4096 bits)
> - Protect private keys with passphrases
> - Disable root login and password authentication on servers
> - Use non-standard ports for public-facing servers
> - Implement fail2ban or similar for brute force protection
> - Regularly update SSH server and client software
> - Rotate SSH keys regularly, especially after team member departures

Reviewer: Let's remove these! You can just click accept suggestions instead of bothering OH on this one :)
```suggestion
```

Copy a file to the remote server:
```bash
scp /path/to/local/file username@hostname:/path/to/remote/directory/
```

Copy a file from the remote server:
```bash
scp username@hostname:/path/to/remote/file /path/to/local/directory/
```

Copy a directory recursively:
```bash
scp -r /path/to/local/directory username@hostname:/path/to/remote/directory/
```

## SSH Agent

Start the SSH agent:
```bash
eval "$(ssh-agent -s)"
```

Add a key to the agent:
```bash
ssh-add ~/.ssh/key_name
```

## Troubleshooting

- Check SSH service status on remote: `systemctl status sshd`
- Verify SSH port is open: `nc -zv hostname 22`
- Debug connection issues: `ssh -vvv username@hostname`
- Check permissions: SSH private keys should have 600 permissions (`chmod 600 ~/.ssh/key_name`)
- Verify known_hosts: If host key changed, remove the old entry with `ssh-keygen -R hostname`

## Secure SSH Key Management

### Storing SSH Keys in AWS Secrets Manager

For secure storage and retrieval of SSH keys across multiple machines, you can use AWS Secrets Manager:

```python
# Store SSH keys in AWS Secrets Manager
python /workspace/store_ssh_keys.py

# Retrieve SSH keys from AWS Secrets Manager

> python /workspace/retrieve_ssh_keys.py

Reviewer: I'd suggest maybe remove this section for AWS, since these scripts won't be always available for agent to use?

Once #8360 goes in, we can just build an AWS microagent with MCP support to fetch the credentials

Author: That's done.

- Verify known_hosts: If host key changed, remove the old entry with `ssh-keygen -R hostname`

## Secure SSH Key Management

### Local Storage with Proper Permissions

The most basic approach is to ensure proper file permissions:

```bash
# Set correct permissions for private keys
chmod 600 ~/.ssh/id_ed25519
# Set correct permissions for public keys
chmod 644 ~/.ssh/id_ed25519.pub
# Set correct permissions for SSH directory
chmod 700 ~/.ssh
```

### Using a Password Manager

Many password managers support storing SSH keys securely:
- 1Password
- LastPass
- Bitwarden
- KeePass

### Using a Secrets Manager

For more advanced key management across multiple machines, consider using a dedicated secrets manager like HashiCorp Vault, Azure Key Vault, or other enterprise solutions.

## Security Best Practices

- Use key-based authentication instead of passwords
- Use strong, unique keys (ED25519 or RSA with at least 4096 bits)
- Protect private keys with passphrases
- Disable root login and password authentication on servers
- Use non-standard ports for public-facing servers
- Implement fail2ban or similar for brute force protection
- Regularly update SSH server and client software

> - Rotate SSH keys regularly, especially after team member departures

Reviewer: These are still not removed :(

Author: Funny enough, I had just asked OpenHands to address the comments.
![image](https://github.com/user-attachments/assets/c85db685-15fa-4f82-94c9-f37930e1c11d)
Hopefully this is better. If not, be explicit please and I'll do it manually.

---
name: SSH Microagent
type: knowledge
version: 1.0.0
agent: CodeActAgent
triggers:
  - ssh
  - remote server
  - remote machine
  - remote host
  - remote connection
  - secure shell
  - ssh keys
---

# SSH Microagent

This microagent provides capabilities for establishing and managing SSH connections to remote machines.

## Capabilities

- Establish SSH connections using password or key-based authentication
- Generate and manage SSH key pairs
- Configure SSH for easier connections
- Execute commands on remote machines
- Transfer files between local and remote machines
- Manage SSH configurations and known hosts
- Store and retrieve SSH keys securely using various methods (local or password managers)

## Authentication Methods

### Password Authentication

```bash
ssh username@hostname
```

> When prompted, the user will need to enter their password.

Reviewer: ```suggestion
When prompted, you should ask the user for their password or a private key.
```

```bash
scp /path/to/local/file username@hostname:/path/to/remote/directory/
```

Copy a file from the remote server:
```bash
scp username@hostname:/path/to/remote/file /path/to/local/directory/
```

Copy a directory recursively:
```bash
scp -r /path/to/local/directory username@hostname:/path/to/remote/directory/
```

## SSH Agent

Start the SSH agent:
```bash
eval "$(ssh-agent -s)"
```

Add a key to the agent:
```bash
ssh-add ~/.ssh/key_name
```

## Troubleshooting

- Check SSH service status on remote: `systemctl status sshd`
- Verify SSH port is open: `nc -zv hostname 22`
- Debug connection issues: `ssh -vvv username@hostname`
- Check permissions: SSH private keys should have 600 permissions (`chmod 600 ~/.ssh/key_name`)
- Verify known_hosts: If host key changed, remove the old entry with `ssh-keygen -R hostname`

## Secure SSH Key Management

### Local Storage with Proper Permissions

The most basic approach is to ensure proper file permissions:

```bash
# Set correct permissions for private keys
chmod 600 ~/.ssh/id_ed25519
# Set correct permissions for public keys
chmod 644 ~/.ssh/id_ed25519.pub
# Set correct permissions for SSH directory
chmod 700 ~/.ssh
```

> ### Using a Password Manager
> 
> Many password managers support storing SSH keys securely:
> - 1Password
> - LastPass
> - Bitwarden
> - KeePass
> 
> ### Using a Secrets Manager
> 
> For more advanced key management across multiple machines, consider using a dedicated secrets manager like HashiCorp Vault, Azure Key Vault, or other enterprise solutions.
> 
> ## Security Best Practices
> 
> - Use key-based authentication instead of passwords
> - Use strong, unique keys (ED25519 or RSA with at least 4096 bits)
> - Protect private keys with passphrases
> - Disable root login and password authentication on servers
> - Use non-standard ports for public-facing servers
> - Implement fail2ban or similar for brute force protection
> - Regularly update SSH server and client software
> - Store SSH keys securely using a secrets manager
> - Rotate SSH keys regularly, especially after team member departures

Reviewer: ```suggestion
```

These seems a bit redundant?
